### PR TITLE
Separate filename and title

### PR DESCRIPTION
**Pull request by natdash**
Hi - this PR is to allow filenames to be separated from titles. 

This allows using ``zetteldeft-id-filename-separator`` and ``deft-file-naming-rules`` to remove whitespace and other characters without affecting other functionality. The documentation still needs to be updated, but I'm hoping that these are changes you'd like to incorporate.

### CODE REVIEW

**Comment by EFLS** [comment]
That does indeed seem like an important feature for some use cases. I'd need to take a closer look at the code, which on first glance goes above my own `elisp` skill level :smile:

Do you think you'd be able to squash all these commits into one? That would make for easier reviewing.

**Comment by EFLS** [comment]
And could this be used to address #31?

**Comment by natdash** [comment]
No problem - I've squashed the commits now.

Re #31 I think it could help. ``deft-file-naming-rules`` takes a function for modifying case, which you might be able to use to strip out unwanted characters. I'll try it out.

**Comment by EFLS** [comment]
Looks promising. I'll try to take a closer look one of these days or weeks, but am a little busy with other things right now.

**Comment by EFLS** [comment]
I'm trying to wrap my head around the lambda's in that final function, but I guess this is where my `elisp` skills fall short. I'll run some tests later to see whether things work as expected, and then include your commits in the `.org` file from which the `.el` is tangled.

**Comment by natdash** [comment]
Which function are you looking at? I can provide an overview if that would be helpful?

**Comment by EFLS** [comment]
That'd be `zetteldeft-file-rename` and `zetteldeft-change-title-and-file-name`.

I'm not sure what lambda's do (never used them before), and haven't seen the `,@` before. What are these called?

**Comment by natdash** [comment]
Ah - they're anonymous functions. The first lambda corresponds to ``cusprompt`` in ``zetteldeft-rename``. It takes the original filename and returns a pair containing the prompt text and the default prompt value. They're used as parameters for ``read-string`` on line 365. 

The ``,@`` marker is documented [here](https://www.gnu.org/software/emacs/manual/html_node/elisp/Backquote.html#Backquote). It evaluates the value and inserts it in a list at the same level. There might be a better way to do this, but it results in something like ``("Rename old-name to (without extension): " "old-name")``.

File: zetteldeft.el [hunk 1]
```diff
@@ -221,10 +228,11 @@ and the file name (without extension) is added to the kill ring.
 When `evil' is loaded, enter insert state."
   (interactive (list (read-string "name: ")))
   (let* ((zdId (zetteldeft-generate-id))
-         (zdName (concat zdId " " str)))
+         (zdName (concat zdId zetteldeft-id-filename-separator str)))
+  (deft-filter str t)
```
Comment (EFLS) [inline zetteldeft.el:232]: What's the reason for adding this line?

Comment (natdash) [reply zetteldeft.el:232]: I must have been looking at deft-new-file . I've removed that line locally and it's working fine for me - will update.

File: zetteldeft.el [hunk 2]
```diff
@@ -326,52 +334,103 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
-    (replace-regexp-in-string
-     (concat zetteldeft-id-regex "[[:space:]]")
-     "" baseName)))
+  (let (contents title)
+    (with-current-buffer (get-buffer-create "*Zetteldeft temp*")
+      (insert-file-contents zdFile nil nil nil t)
+      (setq contents (concat (buffer-string))))
+    (setq title (deft-parse-title zdFile contents))
+    (kill-buffer "*Zetteldeft temp*")
+    title))
```
[the remark this answers is not in the document]
Comment (natdash) [reply zetteldeft.el:355]: That's a fair point - `zetteldeft-new-file` inserts the title though, and `deft` will obtain the title the same way when filtering.

**Comment by natdash** [comment]
I'm using `zetteldeft-change-title-and-file-name` in place of `zetteldeft-file-rename` - if you're happy with the behaviour, I can rename it and remove the original function so that existing keybindings work.

Which function are you seeing the ID issue with? I'll test and resolve.

**Comment by EFLS** [comment]
It happens when I rename a file with `zetteldeft-change-title-and-filename`.

After using it
 - the filename now contains two IDs (where it should be one)
 - the title in the file contains one ID (where it should be none)

 For example: filename `2020-05-21-1549 Testfile.org` becomes `2020-05-21-1549 2020-05-21-1549 Testfile with new title.org`, with contents

    #+title: 2020-05-21-1549 Testfile with new title

**Comment by natdash** [comment]
I haven't been able to replicate this behaviour yet - I've disabled all additional deft and zetteldeft configuration so my filename format is e.g. `2020-05-22-0545 Test note.org`.

Does your prompt display the title only or does it include the ID? `zetteldeft-change-title-and-filename` uses `zetteldeft--lift-file-title` to get the old title - the lambda to generate the new file name only inserts the ID once, and the title should be updated with whatever was returned at the prompt.

What do you see if you evalute `zetteldeft--lift-file-title`? It should behave like this:
```
(zetteldeft--lift-file-title "~/notes/2020-05-22-0545 Test note.org")
"Test note"
```
After re-enabling my configuration, it still works the same way for me:
```
(zetteldeft--lift-file-title "~/notes/2020-05-22-0605_test-note.org")
"Test note"
```

**Comment by EFLS** [comment]
Found it: it depends on the value of `deft-use-filename-as-title`. I have it set to `t` by default (because I like to see the note IDs in the Deft buffer), but it is `nil` by default.

When set back to `nil`, the `zetteldeft--lift-file-title` works as expected.

Perhaps we need to throw in a `(let ((deft-use-filename-as-title nil)))` in that function? Or is there a better way?

**Comment by natdash** [comment]
Thanks! I've implemented the fix and it worked for me. 

What do you think about renaming ``zetteldeft-change-title-and-filename`` to ``zetteldeft-file-rename``?

**Comment by EFLS** [comment]
Please do rename it. I'll then use the changes to see whether it breaks existing workflows.

Can you also share your Deft configuration for altering the filenames? An example would help me write some documentation and maybe add it to the introduction.

**Comment by natdash** [comment]
I've renamed the function now. 

My Deft+zetteldeft config is below:
```
(setq zetteldeft-id-filename-separator "_")
(setq deft-extensions '("org" "md" "txt" "lagda.org"))
(setq deft-default-extension "org")
(setq deft-use-filter-string-for-filename t)
(defun character-filter (str)
  (downcase (replace-regexp-in-string "?" "" str)))
(setq deft-file-naming-rules '((noslash . "-")
			       (nospace . "-")
			       (case-fn . character-filter)))
```
Multiple file extensions are supported as well - the above works for me, but if you have an opportunity to test too. After creating a file, I use ``zetteldeft-file-rename-full`` to modify the extension. Renaming the file later retains the new file extension.

**Comment by jeffvalk** [comment]
I'm very glad to see this PR. It's something that occurred to me as well (see #63). Thanks, @natdash!

From a quick test (definitely not thorough) of 8faaef60bb43847d08a60f2f2c1de1111ada35e4, the code works correctly for my minimal configuration.

I also think there are a few stylistic things that could help the clarity and maintainability of this code:
1. Using conventional elisp indentation throughout;
2. Treating variables in a more functional/lispy (less imperative) way, using serial `let*` binding rather than `setq`; and
3. Putting line breaks in long lines of code.

I'll make some specific review comments on the code, and point things out where they occur. Please know that these thoughts are intended positively and with appreciation. :smile:

File: zetteldeft.el [hunk 3]
```diff
@@ -6,7 +6,7 @@
 ;; URL: https://efls.github.io/zetteldeft/
 ;; Keywords: deft zettelkasten zetteldeft wp files
 ;; Version: 0.2
-;; Package-Requires: ((emacs "25.1") (deft "0.8"))
+;; Package-Requires: ((emacs "25.1") (deft "0.8") (dash "2.17.0"))
```
Comment (jeffvalk) [inline zetteldeft.el:9]: I'm all for using `dash`. That said, is adding a dependency acceptable?

File: zetteldeft.el [hunk 4]
```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
-  "Rename the current file via the deft function.
-Use this on files in the `deft-directory'."
-  (interactive)
+  (let ((deft-use-filename-as-title nil)
+	contents title)
+    (with-current-buffer (get-buffer-create "*Zetteldeft temp*")
+      (insert-file-contents zdFile nil nil nil t)
+      (setq contents (concat (buffer-string))))
+    (setq title (deft-parse-title zdFile contents))
+    (kill-buffer "*Zetteldeft temp*")
```
Comment (jeffvalk) [inline zetteldeft.el:354]: 1. Wouldn't using `with-temp-buffer` be simpler?
2. Creating unassigned variables and then calling `setq` is awkward. Prefer `let*`.
3. The `concat` call is superfluous in `(concat (buffer-string))`.

File: zetteldeft.el [hunk 5]
```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
+    title))
+
+(defun zetteldeft--rename (cusprompt fn-gen update-title)
```
Comment (jeffvalk) [inline zetteldeft.el:357]: This function name should indicate that it both (1) renames the file and (2) updates the title in the buffer.

Also, passing prompt/action functions as arguments to this function is a flexible approach, but merits more commentary to indicate how it's designed and used. I had to read the code more than once to see what was going on here.

File: zetteldeft.el [hunk 6]
```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
-        (zetteldeft-update-title-in-file)
-        (deft-refresh))))
-
-(defun zetteldeft-update-title-in-file ()
+  (let ((old-filename (buffer-file-name))
+	prompt-pars new-name new-filename)
+    (when old-filename
+      (setq prompt-pars (funcall cusprompt old-filename))
+      (setq new-name (read-string (car prompt-pars) (cdr prompt-pars)))
+      (setq new-filename (funcall fn-gen old-filename new-name))
```
Comment (jeffvalk) [inline zetteldeft.el:366]: Creating unassigned variables and then calling `setq` is awkward. Prefer `let*`.

File: zetteldeft.el [hunk 7]
```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
+  (interactive)
+  (zetteldeft--rename
+   (lambda (old-filename) (let (old-title prompt-text)
+			    (setq old-title (zetteldeft--lift-file-title old-filename))
+			    (setq prompt-text (concat "Change " old-title " to: "))
+			    `((,@prompt-text) (,@old-title))))
+   (lambda (old-filename new-name) (let (old-id old-extension new-filename)
+				     (setq old-id (zetteldeft--lift-id old-filename))
+				     (setq old-extension (zetteldeft--identify-extension old-filename))
+				     (setq new-filename (concat (file-name-sans-extension (deft-absolute-filename (concat old-id zetteldeft-id-filename-separator new-name))) "." old-extension))
```
Comment (jeffvalk) [inline zetteldeft.el:394]: 1. Creating unassigned variables and then calling `setq` is awkward. Prefer `let*`.
2. Indentation here doesn't keep with conventional elisp style.
3. Lines are quite long here, which gets hard to read.

File: zetteldeft.el [hunk 8]
```diff
@@ -326,52 +333,89 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
+				     new-filename))
+   t))
+
+(defun zetteldeft--identify-extension (filename)
+  "Identify the extension of the provided FILENAME.
+Handles multiple file extensions, if extension exists in deft-extensions."
+  (let (extension)
+      (setq extension (-reduce-from (lambda (a x) (if (string-match (concat x "$") filename) x a)) nil (--sort (< (length it) (length other)) deft-extensions)))
+      (if extension extension (file-name-extension filename))))
```
Comment (jeffvalk) [inline zetteldeft.el:403]: Is this necessary? This seems to redefine the concept of a file extension, which is traditionally only the final dotted segment. For instance, the file extension for `archive.tar.gz` is `gz`; the string ".tar" is part of the file name.

At the very least, this feature seems unrelated to this PR and better as a separate issue/commit.

File: zetteldeft.el (hunk 3 above)
Comment (natdash) [reply zetteldeft.el:9]: I used `dash` to get `-reduce-from`. Is there an alternative I could use to avoid the dependency?

File: zetteldeft.el (hunk 4 above)
Comment (natdash) [reply zetteldeft.el]: That makes a *lot* of sense - updated!

File: zetteldeft.el (hunk 5 above)
Comment (natdash) [reply zetteldeft.el]: I've updated the function name, and provided a more detailed explanation in the doc string.

File: zetteldeft.el (hunk 6 above)
Comment (natdash) [reply zetteldeft.el]: Thanks - updated!

File: zetteldeft.el (hunk 7 above)
Comment (natdash) [reply zetteldeft.el]: Thanks - updated. I hope the indentation and formatting changes have improved readability.

File: zetteldeft.el (hunk 8 above)
Comment (natdash) [reply zetteldeft.el:408]: My driver to add this feature was to support literate Agda with the file extension `.lagda.org`. The default behaviour is still to only use the final dotted segment, it just allows for a "custom" file extension to be supported if it's added to `deft-extensions`.

**Comment by natdash** [comment]
Thanks very much for the feedback @jeffvalk ! I've implemented several changes from your review so far.

File: zetteldeft.el (hunk 3 above)
Comment (jeffvalk) [reply zetteldeft.el:9]: Packages `cl-lib` and `seq` both have reduce functions, I believe, and are included with emacs.

If there are consistent functional patterns in the code, `dash` is a great choice. If it's just a one-off usage, I'd probably lean toward not adding dependencies.

File: zetteldeft.el (hunk 8 above)
Comment (jeffvalk) [reply zetteldeft.el]: Wouldn't it make more sense to use a supported file extension like `.lagda-org` and associate your desired major mode via `auto-mode-alist` in your `init.el` file? If you try to redefine what a file extension is, it's not just zetteldeft you'd have to change; it's emacs own behavior, and any other package you might install.

**Comment by jeffvalk** [comment]
Sure, happy to add another set of eyes! This was an itch for me too (see #63).

I wonder if it might be possible to make this patch a bit simpler. In particular:
1. I think passing functions as arguments may be over-engineering for the current purpose; and
2. Redefining/implementing file extension handling seems like a symptom of a separate issue (see review comment above).

I'd pulled together some similar code, which I posted [here][1] for reference. Please feel free to borrow and steal any/all of it as you like! :slightly_smiling_face: 

[1]: https://github.com/jeffvalk/zetteldeft/commit/da60afb3d47c2afa22de8aac36e98f4c0b4b523a

**Comment by natdash** [comment]
Thanks again @jeffvalk . I've moved the file rename logic back into `zetteldeft-file-rename` and `zetteldeft-file-rename-full` - agree that it's easier to follow now. I was also able to use `cl-lib` to remove the dependency on `dash`. Are you happy with me incorporating your changes around id generation in `zetteldeft-new file` and the rename functions?

I'd still like to include the custom file extension support if possible, but I understand your point that it may be worth considering as a separate issue. I'm trying to keep the title separate from the filename - a title like `Test note.lagda` would work but then it's not just a title. @EFLS what are your thoughts?

File: zetteldeft.el (hunk 8 above)
Comment (natdash) [reply zetteldeft.el]: I wish it was that easy:
```
test.lagda-org: Unsupported extension.
Supported extensions are: .agda, .lagda, .lagda.rst, .lagda.tex,
                          .lagda.md, .lagda.org
```
The major mode integrates with the compiler, and its literate mode uses those extensions.

Comment (jeffvalk) [reply zetteldeft.el]: Why not use `.lagda` then? It's supported by your compiler and is, without ambiguity, a file extension.

**Comment by jeffvalk** [comment]
Sure, @natdash. Glad to help.

> Are you happy with me incorporating your changes around id generation in zetteldeft-new file and the rename functions?

IMO, pull requests should be narrowly focused and avoid combining topics whenever possible. This makes repository history much easier to understand and maintain. In the current case, I'd say the following three things should be separate issues/PRs:
1. Allowing file name to differ from title (i.e. supporting `deft-file-naming-rules`)
2. Supporting non-standard file extension usage
3. Permitting custom ID generation

Of course, this is up to @EFLS.

File: zetteldeft.el (hunk 8 above)
Comment (natdash) [reply zetteldeft.el]: Sorry, I didn't put enough information in my last post. The `.lagda` extension is for TeX, the only way the compiler appears to support org mode is with `.lagda.org`. I found the source [here](https://github.com/agda/agda/blob/32664c045fffd6708c2d46204a70fcd48f063fe3/src/full/Agda/Syntax/Parser/Literate.hs#L92-L100)

I've had another idea, will post in the main thread below.

**Comment by EFLS** [comment]
I'm a bit busy at the moment, but will hopefully have time to look at the code in a couple of days. I agree with @jeffvalk and would like to implement the changes step by step. So let's use this PR for supporting `deft-file-naming-rules` and separating filename from title. The challenges with other file extensions seem like a separate issue.

**Comment by natdash** [comment]
I'm beginning to agree with @jeffvalk as well. 

What if I revert my last commit, keeping the renaming code in `zetteldeft--rename-file-and-title`, and then remove the custom extension logic from the PR? 

Custom renaming functionality could then be implemented if required in a user's own emacs configuration. I could optionally document my extension handling function as an example in `zetteldeft.org`?

File: zetteldeft.el (hunk 8 above)
Comment (jeffvalk) [reply zetteldeft.el]: Ah, got it. That source clarifies things.

Out of curiosity, have you tried Agda with `org-babel`?

**Comment by jeffvalk** [comment]
The minimal reference [patch][1] I posted above might be a helpful comparator. That example untethers titles from file names, enabling `deft-file-naming-rules` -- and nothing else. If we think this PR should include features beyond that, perhaps it would be helpful to say what and why.

[1]: https://github.com/jeffvalk/zetteldeft/commit/da60afb3d47c2afa22de8aac36e98f4c0b4b523a

**Comment by natdash** [comment]
I think I've made the required changes now - when possible @jeffvalk do you mind having another look?

File: zetteldeft.el [hunk 9]
```diff
@@ -207,7 +207,12 @@ This is done with the regular expression stored in
   (interactive (list
     (completing-read "File to insert full title from: "
       (deft-find-all-files-no-prefix))))
-  (insert (concat zetteldeft-link-indicator (file-name-base file))))
+  (insert (concat zetteldeft-link-indicator (zetteldeft--lift-id file) " " (zetteldeft--lift-file-title (concat deft-directory file)))))
```
[the remark this answers is not in the document]
Comment (natdash) [reply zetteldeft.el]: Thanks - fixed.

File: zetteldeft.el [hunk 10]
```diff
@@ -326,52 +331,68 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
   "Return the title of a zetteldeft note.
 ZDFILE should be a full path to a note."
-  (let ((baseName (file-name-base zdFile)))
-    (replace-regexp-in-string
-     (concat zetteldeft-id-regex "[[:space:]]")
-     "" baseName)))
+  (let ((deft-use-filename-as-title nil))
+    (deft-parse-title zdFile (with-temp-buffer
+      (insert-file-contents zdFile nil nil nil t)
+       (buffer-string)))))
```
[the remark this answers is not in the document]
Comment (natdash) [reply zetteldeft.el]: Updated.

File: zetteldeft.el [hunk 11]
```diff
@@ -326,52 +331,68 @@ Don't forget to add `\\n' at the beginning to start a new line."
...
+  (let ((old-filename (buffer-file-name)))
+    (when old-filename
+      (let* ((old-name (file-name-nondirectory old-filename))
+	     (prompt-text (concat "Rename " old-name " to: "))
+	     (new-name (read-string prompt-text old-name))
+	     (deft-dir (file-name-as-directory deft-directory))
+	     (new-filename (concat deft-dir new-name)))
+	(rename-file old-filename new-filename)
+	(deft-update-visiting-buffers old-filename new-filename)
+	(deft-refresh)))))
```
[the remark this answers is not in the document]
Comment (natdash) [reply zetteldeft.el:400]: I used this function recently to modify some filenames where the ID wasn't unique - I'd created a few notes in the same minute. You're right, `deft-rename-file` would have worked. I like having a function to rename the file in the current buffer instead of from the deft buffer, but this could live in my emacs config. Happy to remove if you think it should go.

[the remark this answers is not in the document]
Comment (jeffvalk) [reply zetteldeft.el:400]: I do think this belongs in your config rather than in zetteldeft, yes.

**Comment by jeffvalk** [comment]
@EFLS: Since some of @natdash's motivations for file renaming relate to having non-unique IDs when creating more than one note per minute, might it make sense to do more to prevent this? Currently, this is mentioned in the code commentary for `zetteldeft-id-format` [here][1]. I wonder if this might be something to mention in the introduction/setup documentation...or if ID generation should prevent this.

This is a separate topic from this PR, but seemed worth asking.

[1]: https://www.eliasstorms.net/zetteldeft/zetteldeft.html#org5f33ae1

**Comment by jeffvalk** [comment]
@natdash To make this PR mergeable, would you please pull the upstream changes into your branch?

**Comment by natdash** [comment]
I've pulled in the upstream changes now, and removed `zetteldeft-file-rename-full`.

**Comment by EFLS** [comment]
This is looking great! Thanks for merging with the latest version, and including literate code in the `.org`. I'll take a look at the final changes over the weekend.

**Comment by EFLS** [comment]
Went through the changes and seems to work fine. One issue though, is that when I use `zetteldeft-find-file-full-title-insert`, I get an error because the file's extension isn't included in the full path...

Do you get a similar issue, @natdash?

**Comment by EFLS** [comment]
Never mind, found it. Minor issue in said function. `zetteldeft--lift-file-title` should be called on the full filename (rather than on its `file-name-base`).

**Comment by EFLS** [comment]
There's still one issue I run into. When using `zetteldeft-find-file-full-title-insert` I get an error. I think because the new `zetteldeft--lift-file-title` expects a full path but is only provided a relative filename.

**Comment by EFLS** [comment]
> @EFLS: Since some of @natdash's motivations for file renaming relate to having non-unique IDs when creating more than one note per minute, might it make sense to do more to prevent this? Currently, this is mentioned in the code commentary for `zetteldeft-id-format` [here](https://www.eliasstorms.net/zetteldeft/zetteldeft.html#org5f33ae1). I wonder if this might be something to mention in the introduction/setup documentation...or if ID generation should prevent this.

Should definitely be prevented, but not sure what the best approach is here. A check and a warning are a minimum. In any case, a separate issue indeed.

**Comment by EFLS** [comment]
I've now also tested it with some deft naming rules: 

```` emacs-lisp
(setq deft-file-naming-rules
  '((noslash . "-")
    (nospace . "-")
    (case-fn . downcase)))
````

New files already replace all spaces with `-`, so I'm wondering whether the `zetteldeft-id-filename-separator` as customizable setting is really necessary? The default space will be replaced anyway, or am I missing something?

**Comment by jeffvalk** [comment]
> I'm wondering whether the `zetteldeft-id-filename-separator` as customizable setting is really necessary?

Stylistically, I tend to think it's better to declare the separator as a variable than to hard code it in multiple places.

Whether the value should be customizable (`defcustom` vs `defvar`) is an separate question. I don't see an obvious use case for customization, but allowing it doesn't seem harmful.

**Comment by natdash** [comment]
I agree - not sure if there's any real benefit there but it doesn't hurt to allow the user to configure the value. Maybe someone would prefer to use an underscore to replace spaces and slashes with `deft-file-naming-rules`, but a hyphen between id and formatted title.

**Comment by jeffvalk** [comment]
Sounds good to me. Is there anything remaining on this PR?

**Comment by EFLS** [comment]
All makes sense! Thanks for your valuable contributions.
